### PR TITLE
Support Snappy and Zstd for confighttp.go

### PR DESCRIPTION
**Pull request by hyunuk**
**Description:**
This PR provides [snappy](github.com/golang/snappy) and [zstd](https://github.com/klauspost/compress/tree/master/zstd) as additional compression methods for `confighttp`, which are the same libraries as `configgrpc` uses.

The `compression` field in `exporter/otlphttpexporter/config.go` has been removed and integrated into `confighttp.HTTPClientSettings`.

**Link to tracking Issue:**
#2766

### CODE REVIEW

**Comment by alolita** [comment]
@open-telemetry/collector-approvers please review.

File: internal/middleware/compression.go
Comment (jpkrohling) [inline internal/middleware/compression.go:48]: You can also store just a reference to the writer function. Like:

```golang
type CompressRoundTripper struct {
	RoundTripper    http.RoundTripper
	CompressionType string
	Writer          func(*bytes.Buffer) (io.WriteCloser, error)
}
```

And then:
```golang
func NewCompressRoundTripper(rt http.RoundTripper, compressionType string) *CompressRoundTripper {
	return &CompressRoundTripper{
		RoundTripper:    rt,
		CompressionType: compressionType,
		Writer:          writerFactory(compressionType),
	}
}
```

And then:

```golang
func writerFactory(compressionType string) func(*bytes.Buffer) (io.WriteCloser, error) {
	switch compressionType {
	case CompressionGzip:
		return func(buf *bytes.Buffer) (io.WriteCloser, error) {
			return gzip.NewWriter(buf), nil
		}
	default:
		return nil
	}
}
```

Comment (jpkrohling) [reply internal/middleware/compression.go:48]: Note: you probably want to have the call to `writerFactory` to happen in a separate statement and check for nil, indicating that the compression type is not known.

File: config/confighttp/confighttp.go
Comment (hyunuk) [reply config/confighttp/confighttp.go:86]: done

File: config/confighttp/confighttp.go
Comment (hyunuk) [reply config/confighttp/confighttp.go:116]: validate at here to avoid additional calls.

File: internal/middleware/compression.go
Comment (hyunuk) [reply internal/middleware/compression.go:44]: created a validation function at confighttp.go, so `compressionType` in compression.go ensures no error.

File: internal/middleware/compression.go
Comment (hyunuk) [reply internal/middleware/compression.go:47]: done. Thanks for letting me know.

File: internal/middleware/compression.go
Comment (hyunuk) [reply internal/middleware/compression.go]: done

File: config/confighttp/confighttp.go
Comment (codeboten) [inline config/confighttp/confighttp.go:115]: should this include `deflate` as well in this list?

File: exporter/otlphttpexporter/README.md
Comment (codeboten) [inline exporter/otlphttpexporter/README.md:36]: Is this needed here as well as in the confighttp readme? For configgrpc is only in one place.

File: CHANGELOG.md
Comment (codeboten) [inline CHANGELOG.md:7]: Please add a note for users to get some guidance about the configuration change needed to move the compression configuration out of the exporter config and into the http client config.

File: config/confighttp/confighttp_test.go
Comment (codeboten) [inline config/confighttp/confighttp_test.go:98]: It looks like the test for no compression was lost in this change, is it still relevant?

Comment (hyunuk) [reply config/confighttp/confighttp_test.go:98]: test for no compression is treated as uncompressed, and it's included in the first test case, "all_valid_settings".

File: config/confighttp/confighttp.go
Comment (hyunuk) [reply config/confighttp/confighttp.go]: fixed

File: exporter/otlphttpexporter/README.md
Comment (hyunuk) [reply exporter/otlphttpexporter/README.md]: deleted

File: CHANGELOG.md
Comment (hyunuk) [reply CHANGELOG.md]: added. Thank you for letting me know

File: config/confighttp/confighttp.go
Comment (Aneurysm9) [inline config/confighttp/confighttp.go:55]: This isn't necessary, is it?  Since the underlying type of `CompressionType` is `string` it should be possible to use a `CompressionType` value as a string.  If it can't be used directly then it can be coerced with `string(val)`.  It looks like this map is also used for validation, but that can be done with a set:

```go
validCompressionKeys = map[CompressionType]struct{}{
  CompressionGzip: {},
...
}
```

Then for validation just check to see if the entry is in the set:

```go
_, valid := validCompressionKeys[CompressionType(strVal)]
```

File: config/confighttp/confighttp.go
Comment (Aneurysm9) [inline config/confighttp/confighttp.go:120]: Should this return an error, or simply not use compression?  Both outcomes could be surprising and I'm not sure which is the right choice.

File: config/confighttp/confighttp.go
Comment (Aneurysm9) [inline config/confighttp/confighttp.go:39]:
```suggestion
	CompressionUnsupported CompressionType = "<unsupported>"
```

I would use a value here that is unlikely to appear accidentally.

File: config/confighttp/confighttp.go
Comment (Aneurysm9) [inline config/confighttp/confighttp.go:44]:
```suggestion
	CompressionZstd        CompressionType = "zstd"
	CompressionNone      CompressionType = "none"
```

There should be an explicit option for specifying no compression.

File: internal/middleware/compression.go
Comment (Aneurysm9) [inline internal/middleware/compression.go:41]: Do these fields need to be exported?  It seems unwise to allow them to be changed after creation.

```suggestion
	compressionType string
	writer          func(*bytes.Buffer) (io.WriteCloser, error)
```

File: config/confighttp/confighttp.go
Comment (jpkrohling) [reply config/confighttp/confighttp.go:120]: The choice of compression is an explicit one, and not doing what the user is explicitly requesting would be an error, IMO.

File: config/confighttp/README.md
Comment (jpkrohling) [inline config/confighttp/README.md:23]: Should we be making a recommendation here to users? According to the notes from the SIG meeting from Nov 3rd, `zstd` would be the 1st option, with `snappy` being the 2nd.

File: config/confighttp/README.md
Comment (jpkrohling) [inline config/confighttp/README.md:38]: If we are making a recommendation, the value from this example should follow the recommendation.

File: config/confighttp/confighttp.go
Comment (jpkrohling) [inline config/confighttp/confighttp.go:118]: If the suggestion about the compression type `none` is accepted, you'll need to handle this special case as well.

File: internal/middleware/compression.go
Comment (jpkrohling) [inline internal/middleware/compression.go:70]: I guess the "none" type would fall here if not handled properly.

**Review by jpkrohling — COMMENTED** [review]
This is looking great. I think @Aneurysm9's questions are the last remaining points.

File: config/confighttp/confighttp.go
Comment (hyunuk) [reply config/confighttp/confighttp.go]: done

File: config/confighttp/confighttp.go
Comment (hyunuk) [reply config/confighttp/confighttp.go:120]: I agree with @jpkrohling

File: config/confighttp/confighttp.go
Comment (hyunuk) [reply config/confighttp/confighttp.go]: deleted the usage of `CompressionUnsupported`

File: config/confighttp/confighttp.go
Comment (hyunuk) [reply config/confighttp/confighttp.go]: done

File: internal/middleware/compression.go
Comment (hyunuk) [reply internal/middleware/compression.go]: done

File: config/confighttp/README.md
Comment (hyunuk) [reply config/confighttp/README.md]: changed the order and added recommendation.

File: config/confighttp/README.md
Comment (hyunuk) [reply config/confighttp/README.md]: done

File: config/confighttp/confighttp.go
Comment (hyunuk) [reply config/confighttp/confighttp.go]: changed

File: CHANGELOG.md
Comment (jpkrohling) [inline CHANGELOG.md:6]: I don't think this is a breaking change, is it? I believe this should go into "Enhancements". Also, a suggestion for this entry would be:

```suggestion
- `confighttp`: add client-side compression support.  (#4441)
```

Comment (hyunuk) [reply CHANGELOG.md]: fixed. I thought removing compression from otlphttpexporter will be a breaking change.

File: config/confighttp/confighttp.go
Comment (jpkrohling) [reply config/confighttp/confighttp.go:86]: This can still be

    Compression CompressionType `mapstructure:"compression"`

File: config/confighttp/confighttp.go
Comment (jpkrohling) [inline config/confighttp/confighttp.go:265]: Not sure what's better: the map, or a switch case here, like:

```go
func isValidHTTPCompressionKey(compressionType CompressionType) bool {
	switch compressionType {
	case CompressionDeflate,
		CompressionGzip,
		CompressionSnappy,
		CompressionZlib,
		CompressionNone,
		CompressionZstd:
		return true
	default:
		return false
	}
}
```

Even better would be if mapstruct would support Enums. Do you know whether they do?

Comment (hyunuk) [reply config/confighttp/confighttp.go]: changed to switch case

**Review by jpkrohling — APPROVED** [review]
LGTM

**Review by Aneurysm9 — APPROVED** [review]

**Review by codeboten — APPROVED** [review]

File: config/confighttp/README.md
Comment (bogdandrutu) [inline config/confighttp/README.md:25]: Why zstd and snappy  are recommended? Do we really need to make a recommendation? Why not the most common one gzip?

File: internal/middleware/compression.go
Comment (bogdandrutu) [inline internal/middleware/compression.go:34]: Can these values be shared between gRPC and HTTP? I remember to see them there as well.

File: config/confighttp/README.md
Comment (bogdandrutu) [inline config/confighttp/README.md:23]: I think we decided on a shorter list to support in gRPC why do we extend here?

Comment (bogdandrutu) [reply config/confighttp/README.md:23]: https://github.com/open-telemetry/opentelemetry-collector/blob/main/config/configgrpc/configgrpc.go#L50

File: config/confighttp/README.md
Comment (hyunuk) [reply config/confighttp/README.md:25]: This change was based on @jpkrohling's [comment](https://github.com/open-telemetry/opentelemetry-collector/pull/4441#pullrequestreview-814747382):

> Should we be making a recommendation here to users? According to the notes from the SIG meeting from Nov 3rd, zstd would be the 1st option, with snappy being the 2nd.

File: internal/middleware/compression.go
Comment (hyunuk) [reply internal/middleware/compression.go]: It can be shareable. I removed usages of these values in confighttp and made connection to middleware.CompressionType.
To connect these values with configgrpc, I discussed with @Aneurysm9 and since the changes in confighttp are quite a lot, he will create an issue about refactoring configgrpc after this PR is being merged.

File: config/confighttp/README.md
Comment (hyunuk) [reply config/confighttp/README.md:23]: `zlib` and `deflate` were already supported in otlphttpexporter. I kept it after discussed with @Aneurysm9.

File: config/confighttp/README.md
Comment (jpkrohling) [reply config/confighttp/README.md:25]: As a user of this, I would appreciate being given a hint on what's the current best/preferred algorithm. I don't know why zstd and snappy are recommended, those were the ones mentioned as being "the first choice and second choice" during the SIG meeting where we discussed this, according to the meeting minutes.

Comment (bogdandrutu) [reply config/confighttp/README.md:25]: @jpkrohling that is true they were the first and second to add on top of "gzip". So I think gzip is actually the top.

File: internal/middleware/compression_test.go
Comment (bogdandrutu) [inline internal/middleware/compression_test.go:121]: CompressionNone?

File: config/confighttp/confighttp.go
Comment (bogdandrutu) [inline config/confighttp/confighttp.go:63]: To fail fast (during config unmarshaling) make CompressionType to implement this interface https://pkg.go.dev/encoding#TextUnmarshaler

File: config/confighttp/README.md
Comment (hyunuk) [reply config/confighttp/README.md:25]: As per this issue #2548,
> Gzip is CPU intensive and offers less of a compression benefit than more modern compression like snappy or zstd

I think the reason we introduced `snappy` and `zstd` is due to this downside of `gzip`.

File: config/confighttp/confighttp.go
Comment (hyunuk) [reply config/confighttp/confighttp.go:63]: Thank you for the reply. Fixed it.

File: internal/middleware/compression_test.go
Comment (hyunuk) [reply internal/middleware/compression_test.go]: fixed

File: config/confighttp/README.md
Comment (hyunuk) [reply config/confighttp/README.md:25]: I will change the recommendation part to 
> look at the documentation for the server-side of the communication

As we discussed in the SIG meeting on Dec 8

**Review by bogdandrutu — CHANGES_REQUESTED** [review]
Missing followup from SIG discussion.

**Review by bogdandrutu — APPROVED** [review]

**Pull request merged**
